Adding Docs for the FIRST Driver Station (#3346)
wpilibsuite/frc-docs: source/docs/zero-to-robot/step-2/first-driver-station.rst

# Installing the FIRST Driver Station

This guide will walk you though installing the FIRST Driver Station on your computer.

## Downloading Install Files

Navigate to the [First Drive Station Releases Page](https://github.com/wpilibsuite/FirstDriverStation-Public/releases/latest). On macOS platforms download the correct ``.pkg`` file and on Windows platforms download the correct ``.exe`` file for your device. For Linux devices, only the archive files are provided.

## Per Platform Setup

### Windows

On Windows, the app has everything configured by default. Just run the installer, and then run the application.

### macOS

There are 3 permissions that macOS requires. It requires Input Monitoring, Local Network access, and data access from other apps. You will get popups for these, and they must be accepted in order to work. The first time you start up the app, these prompts will cause the launch to fail, and you'll need to accept the Input Monitoring prompt, and then restart the app. Then you'll be able to accept the Local Network permission.

> If Local Network access is declined, the app will still seem to function normally, as Apple does not provide a way to detect if the permission has been granted.

Reviewer: What is the symptom that users would see if local network access is declined. I assume it won't connect to the robot?

macOS will also prompt with "FirstDriverStation" would like to access data from other apps. This permission is required for launching dashboards and for log file writing. You should click Allow when this prompt appears.

If you decline any of these, you can fix the settings in the ``Privacy & Security`` tab of System Settings.

If Local Network access still does not work after re-enabling it there, see :ref:`macOS Permissions<docs/zero-to-robot/step-2/first-driver-station:macOs Permissions>` for a terminal-based workaround.

### Linux

> 

Reviewer: Add a warning about root or sudo access being required for installation, since normally our tools do not need it

The following packages must be installed in order for linux to work:

.. todo:: update when the FirstDriverStation docs update

Additionally, the app needs to be a part of the input group in order to have input access. That can be done with the following commands.

> ```text

Reviewer: ```suggestion
```bash
```

sudo chgrp input FirstDriverStation
sudo chmod g+s FirstDriverStation
```

Finally, for proper controller access, the current user needs access to hidraw. To do that, create a ``/etc/udev/rules.d/72-hidraw.rules`` containing

> ```text

Reviewer: ```suggestion
```bash
```

```

Then reload udev rules:

```text
sudo udevadm control --reload-rules && sudo udevadm trigger
```

## macOs Permissions

masOs requires both Input Monitoring and Local Network access for the Driver Station to function correctly. If you decline one of these prompt, first try re-enabling access in System Settings->Privacy & Security. If the Driver Station still cannot access the local network, you can add macOS local network exceptions from Terminal for both Ethernet and Wi-Fi. The following commands allow access to any ``10.x.x.x`` address and any ``172.16.x.x`` through ``172.31.x.x`` address:

```text
sudo defaults write com.apple.network.local-network AllowedEthernetLocalNetworkAddresses -array "10.0.0.0/8"
sudo defaults write com.apple.network.local-network AllowedEthernetLocalNetworkAddresses -array-add "172.16.0.0/12"
sudo defaults write com.apple.network.local-network AllowedWiFiLocalNetworkAddresses -array "10.0.0.0/8"
sudo defaults write com.apple.network.local-network AllowedWiFiLocalNetworkAddresses -array-add "172.16.0.0/12"
```

After running these commands, reboot macOS before starting the Driver Station again.

> .. warning:: These settings allow any app on the Mac to access local networks in those ranges without prompting. If you only need access to one robot network, a narrower CIDR range is more restrictive.

Reviewer: I think this update makes it more clear for someone who isn't familiar with networking (borrowed from the VS Code known issue)

```suggestion
.. warning:: This also allows any program to access any network with the address in the range of ``10.x.x.x``, which may be undesirable if you connect to other networks in that range beside's the robot's network. If you only access a robot with a single team number, you can substitute ``10.TE.AM.0/24``` (:ref:`TE.AM IP Notation <docs/networking/networking-introduction/ip-configurations:TE.AM IP Address Notation>`). If you only access the robot's network over Ethernet or WiFi, you can only run the appropriate command.
```
